Import coverage generation module from Zephyr

nmeum opened the pull request:
### Contribution description

This PR imports a slightly modified version of the coverage generation code from [Zephyr](https:/zephyrproject.org/) and allows generating coverage information on real hardware (not just native) for GNU Gcov. The imported `sys/coverage` module writes coverage information on `exit()` to the UART. The output generated by `sys/coverage` can then be parsed using the `dist/tools/coverage/coverage.py` script which generates `.gcda` files from the output. These files can then be passed to Gcov frontends like [gcovr](http://gcovr.com/) or [lcov](https://github.com/linux-test-project/lcov) to generate HTML coverage visualizations.

See also:

* https://docs.zephyrproject.org/1.14.1/guides/coverage.html
* https://github.com/zephyrproject-rtos/zephyr/tree/main/subsys/testsuite/coverage
* https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/gen_gcov_files.py

### Testing procedure

A simple test can be performed with a `hello-world` application modified as follows:

```diff
diff --git a/examples/hello-world/Makefile b/examples/hello-world/Makefile
index 258d8e9baf..944f944d22 100644
--- a/examples/hello-world/Makefile
+++ b/examples/hello-world/Makefile
@@ -15,4 +15,6 @@ DEVELHELP ?= 1
 # Change this to 0 show compiler invocation lines by default:
 QUIET ?= 1
 
+USEMODULE += coverage
+
 include $(RIOTBASE)/Makefile.include
diff --git a/examples/hello-world/main.c b/examples/hello-world/main.c
index f51bf8c0a0..77c19d2c7e 100644
--- a/examples/hello-world/main.c
+++ b/examples/hello-world/main.c
@@ -20,6 +20,7 @@
  */
 
 #include <stdio.h>
+#include <stdlib.h>
 
 int main(void)
 {
@@ -28,5 +29,6 @@ int main(void)
     printf("You are running RIOT on a(n) %s board.\n", RIOT_BOARD);
     printf("This board features a(n) %s MCU.\n", RIOT_MCU);
 
+    exit(0);
     return 0;
 }
```

Compile the application as follows:

```
$ BOARD=nucleo-f401re make -C examples/hello-world/
```

Flash the application using:

```
$ BOARD=nucleo-f401re make -C examples/hello-world/ flash
```

Afterwards wait a few seconds and run `make term` as follows:

```
$ BOARD=nucleo-f401re make -C examples/hello-world/ term | tee /tmp/log.txt
```

Reset the board and wait until `GCOV_COVERAGE_DUMP_END` appears in the output, close the terminal as soon as it does and run the following command:

```
$ ./dist/tools/coverage/coverage.py -i /tmp/log.txt
```

If you see errors about invalid UTF-8 sequences, check the `log.txt` and remove these manually for now and re-run the script (see To-Do below). If the script terminates successfully, this should cause various `.gcda` files to be generated (`find . -name '*.gcda'`). These files can then be passed to a Gcov frontend (e.g. `gcovr`) to generate HTML files:

```
$ gcovr -r . --html -o coverage.html --html-details
```

Finally open `coverage.html` in a web browser.

![riot-coverage-screenshot](https://user-images.githubusercontent.com/2326560/137708544-25728c53-8dfa-4e2b-a427-91ed1472be62.png)

### To-Do

I think the tooling could be improved significantly. For example, a `make coverage` target which works similar to `make term` but invokes the `./dist/tools/coverage/coverage.py` script directly and terminates as soon as it sees `GCOV_COVERAGE_DUMP_END` would be nice to have IMHO. Furthermore, I occasionally noticed UTF-8 decode errors when using the python script since resetting the board sometimes causes invalid UTF-8 sequences to be generated. This should probably be addressed as well. **In general, I haven't tested this a lot yet and would welcome general feedback on the idea and the integration.**

miri64 commented:
Would it make sense to provide a `make coverage` target, that 
1. does everything (or most of) what you describe in the Testing procedure section
2. downloads the sources as a package from the Zephyr repo
?

BTW, is it possible to combine the results of the coverage tests? If so we could integrate that into our CI in the future, maybe.

miri64 commented:
>  For example, a `make coverage` target which works similar to `make term` but invokes the `./dist/tools/coverage/coverage.py` script directly and terminates as soon as it sees `GCOV_COVERAGE_DUMP_END` would be nice to have IMHO.

I guess that would then be `make coverage-term` with my proposal.

miri64 commented:
> > BTW, is it possible to combine the results of the coverage tests? If so we could integrate that into our CI in the future, maybe.
> 
> What do you mean by combine?

E.g. if I run `examples/gnrc_border_router` and `fuzzing/gnrc_tcp`, will the results be combined when piping to ` /tmp/log.txt` or will the `coverage.py` script be just confused in that case. I am asking, since we most likely will never have a one-fits-all application. `tests/unittests` used to be a try on that, but as we know, that has its own (size) problems. And this does not even consider different execution paths due to module selection and `#ifdef`s.

miri64 commented:
Ping @nmeum?

miri64 commented:
> Sorry, I didn't have the time to work further on this so far. However, last time I used this it did actually work fine. Is there anything that I need to do apart from rebasing and resolving merge conflicts? :)

From https://github.com/RIOT-OS/RIOT/pull/17000#issuecomment-948340518

> It should theoretically be possible to ship the coverage generation code from Zephyr in `./pkg` instead of `./sys` with custom RIOT patches on top. However, I am not sure what one would use as a `$PKG_URL` then. Cloning the entire zephyr repository for two files seems a bit overkill. I am not sure if there is any better way to checkout the source with the RIOT package system.

and then the conclusion https://github.com/RIOT-OS/RIOT/pull/17000#issuecomment-948635291

> > Indeed, but given that Zephyr did not put it in its own repository, that seems indeed to be overkill... How long does the fetch take with `--depth=1` roughly? Maybe that's enough. What do you think @fjmolinas and @jia200x?
> 
> just copy those files over

Not sure if this was addressed. Other than that, I guess the dedicated make target and the merging of coverage data from several tests can be addressed in a separate PR, once this is in place.

miri64 commented:
Oh, now after looking at the code again, maybe it would make sense to instead of adding the dumping to the `exit()` function would be to add it to `main_trampoline()` in `core/lib/init.c` instead. This way, the code of an application does not need to be touched to include the coverage dump.

miri64 commented on dist/tools/coverage/coverage.py:
```diff
@@ -0,0 +1,85 @@
+    extracted_coverage_info = {}
+    capture_data = False
+    reached_end = False
+    with open(input_file, 'r') as fp:
+        for line in fp.readlines():
```
```suggestion
    with open(input_file, 'r') as fp:
        for line_bytes in fp.readlines():
            line = line_bytes.decode(errors="ignore")
```

should get rid of the UTF-8 decoding errors you reported. If the UTF-8 characters are not in fact invalid, one could always use another [error handler](https://docs.python.org/3/library/codecs.html#error-handlers).

nmeum commented:
I rebased this, updated the `coverage.c` file from Zephyr upstream, and implemented Miri's trampoline suggestion. It still works in a virtual RISC-V environment. I still need to test this on real hardware again and in this regard I also want to improve integration with the Makefiles, e.g. ideally over a `make coverage` target.

nmeum commented:
I tested this with `BOARD=nucleo-f401re` again and it seems to works as expected:

```
$ git diff examples/hello-world
diff --git a/examples/hello-world/Makefile b/examples/hello-world/Makefile
index 258d8e9baf..aa2c6027dc 100644
--- a/examples/hello-world/Makefile
+++ b/examples/hello-world/Makefile
@@ -12,6 +12,8 @@ RIOTBASE ?= $(CURDIR)/../..
 # development process:
 DEVELHELP ?= 1

+USEMODULE += coverage
+
 # Change this to 0 show compiler invocation lines by default:
 QUIET ?= 1
$ BOARD=nucleo-f401re make -C examples/hello-world/ flash
$ BOARD=nucleo-f401re make -C examples/hello-world/ cleanterm 2>&1 | tee /tmp/out
<reset the board>
$ ./dist/tools/coverage/coverage.py -i /tmp/out
$ gcovr --gcov-executable arm-none-eabi-gcov -r . -s | tail -6
------------------------------------------------------------------------------
TOTAL                                       1887     576    30%
------------------------------------------------------------------------------
lines: 30.5% (576 out of 1887)
functions: 26.5% (54 out of 204)
branches: 16.9% (117 out of 694)
```

A `make coverage` target would probably be nice to have as well (which at least generates the `gcov` files). However, this would probably require writing an additional pyserial Python script which reads data from the UART until it sees `^GCOV_COVERAGE_DUMP_END$` and then passes the retrieved data to `./dist/tools/coverage/coverage.py`. However, this script would also need to deal with all kinds of quirks then. For example, if you reset your device while it is dumping coverage data then you might have two `^GCOV_COVERAGE_DUMP_START$` in your UART dump etc.

Maybe it makes sense to consider merging this without a `make coverage` target for now? The process for obtaining coverage information is documented in `./dist/tools/coverage/README.md`. I would also like to point out that Zephyr also requires manually running the commands outlined above [\[1\]](https://docs.zephyrproject.org/1.14.1/guides/coverage.html#steps-to-generate-code-coverage-reports).

miri64 commented:
> However, this would probably require writing an additional pyserial Python script which reads data from the UART until it sees `^GCOV_COVERAGE_DUMP_END$` and then passes the retrieved data to `./dist/tools/coverage/coverage.py`. However, this script would also need to deal with all kinds of quirks then. For example, if you reset your device while it is dumping coverage data then you might have two `^GCOV_COVERAGE_DUMP_START$` in your UART dump etc.

Mhhh are you sure pyserial is required for that? Sounds complicated and would make it quite incompatible with e.g. `native` why not use exactly what you did in your test output there in the `coverage` target, just without the file indirection oO (assuming reading from `stdin` is supported by the script, if not, why isn't it ;-P)?

```mk
coverage: $(TERMDEPS)
   ${MAKE} -C examples/hello-world/ cleanterm 2>&1 | ./dist/tools/coverage/coverage.py -i -
   ${GCOVR} --gcov-executable arm-none-eabi-gcov -r . -s | tail -6
```

nmeum commented:
Unless I am missing something obvious the problem with `make cleanterm | ./dist/tools/coverage/coverage.py -i -` should be that the first command (`make cleanterm`) doesn't terminate so the command just hangs forever. Hence my suggestion to have a pyserial script which terminates after seeing `^GCOV_COVERAGE_DUMP_END$` on the UART.

nmeum commented:
I added a `--exit-on-match` flag to `pyterm` which is similar to `/trigger` but actually forces termination via `os._exit`. Based on this flag, I added a `make coverage` target which is similar to what you are proposing but doesn't call gcovr automatically (since it is just one out of many gcov frontends).

`make coverage` still requires the board to be reset by the user, not sure if that could be automated as well via OpenOCD.

nmeum commented on dist/tools/pyterm/pyterm:
```diff
@@ -653,6 +657,10 @@ class SerCmd(cmd.Cmd):
 
+        if self.line_regex and self.line_regex.match(line):
+            self.do_PYTERM_exit("")
+            os._exit(0)
```
This is a bit ugly, but otherwise one would need to terminate the cmdloop thread somehow since this code is not executed on the main thread.

nmeum commented:
* Rebased against current master
* Fixed a few bugs
* Improved the documentation and moved it to Doxygen

I also tested the current version again using `BOARD=nucleo-f401re`.

With this board, I was able to successfully generate coverage for `examples/hello_world` with `gcovr`.

![Screenshot 2023-01-16 at 12-35-06 GCC Code Coverage Report](https://user-images.githubusercontent.com/2326560/212669179-6af95f28-ac0c-4fc7-a461-da193cb07ded.png)

@miri64 could you take another look and maybe test if it also works for you? :)

nmeum commented on doc/doxygen/src/debugging-aids.md:
```diff
@@ -1,5 +1,33 @@
 # Debugging Tools {#debugging-tools}
```
Not sure if this is the best subsection for the coverage documentation.

Maybe it makes sense to rename this to `Developer Tooling` or something along those lines?

benpicco commented on sys/include/coverage.h:
```diff
@@ -0,0 +1,118 @@
+ * of the object file containing the selected comdat function.
+ */
+struct gcov_fn_info {
+    const struct gcov_info *key;     /* comdat key */
+    unsigned int ident;              /* unique ident of function */
+    unsigned int lineno_checksum;    /* function lineo_checksum */
+    unsigned int cfg_checksum;       /* function cfg checksum */
+    struct gcov_ctr_info ctrs[0];    /* instrumented counters */
```
```suggestion
    const struct gcov_info *key;     /**< comdat key */
    unsigned int ident;              /**< unique ident of function */
    unsigned int lineno_checksum;    /**< function lineo_checksum */
    unsigned int cfg_checksum;       /**< function cfg checksum */
    struct gcov_ctr_info ctrs[0];    /**< instrumented counters */
```

to make Doxygen happy

benpicco commented on sys/include/coverage.h:
```diff
@@ -0,0 +1,118 @@
+ * at run-time with the exception of the next pointer.
+ */
+struct gcov_info {
+    unsigned int version;        /* Gcov version (same as GCC version) */
+    struct gcov_info *next;      /* List head for a singly-linked list */
+    unsigned int stamp;          /* Uniquifying time stamp */
+#ifdef GCOV_12_FORMAT
+    unsigned int checksum;		/* unique object checksum */
+#endif
+    const char *filename;        /* Name of the associated gcda data file */
+    /* merge functions, null for unused*/
+    void (*merge[GCOV_COUNTERS])(gcov_type *, unsigned int);
+    unsigned int n_functions;         /* number of instrumented functions */
+    struct gcov_fn_info **functions;  /* function information */
+};
```
```suggestion
    unsigned int version;        /**< Gcov version (same as GCC version) */
    struct gcov_info *next;      /**< List head for a singly-linked list */
    unsigned int stamp;          /**< Uniquifying time stamp */
#ifdef GCOV_12_FORMAT
    unsigned int checksum;		/**< unique object checksum */
#endif
    const char *filename;        /**< Name of the associated gcda data file */
    /* merge functions, null for unused*/
    void (*merge[GCOV_COUNTERS])(gcov_type **<, unsigned int);
    unsigned int n_functions;         /**< number of instrumented functions */
    struct gcov_fn_info **functions;  /**< function information */
};
```

benpicco reviewed (commented):
I like how little changes this needs in the code base.
Just some CI nits and a rebase is needed, I think this is then good to merge.

nmeum commented:
Applied @benpicco's suggestions (thanks!) and rebased against master.

benpicco commented:
CI still has [come complaints](https://github.com/RIOT-OS/RIOT/actions/runs/3996810704/jobs/6857317249) about formatting.

benpicco commented:
Please squash

benpicco commented on sys/include/coverage.h:
```diff
@@ -0,0 +1,125 @@
+    unsigned int ident;              /**< unique ident of function */
+    unsigned int lineno_checksum;    /**< function lineo_checksum */
+    unsigned int cfg_checksum;       /**< function cfg checksum */
+    struct gcov_ctr_info ctrs[0];    /**< instrumented counters */
```
sys/include/coverage.h:100:26: error: ISO C forbids zero-size array ‘ctrs’ [-Werror=pedantic]
      100 |     struct gcov_ctr_info ctrs[0];    /**< instrumented counters */

benpicco commented:
Please add a test application that makes use of the `coverage` module so this gets build by CI.
I already fail to compile the example on `native`.

If the call to `exit(0)` is mandatory, would it make sense to use [`test_utils_main_exit_cb()`](https://github.com/RIOT-OS/RIOT/blob/master/core/lib/init.c#L65) for this?

It would only work if the test is run in the main thread, but I think that's the case for the kind of coverage reports we could create here.

nmeum commented:
> Please add a test application that makes use of the `coverage` module so this gets build by CI.

Good idea, will do so later.

> If the call to `exit(0)` is mandatory, would it make sense to use [`test_utils_main_exit_cb()`](https://github.com/RIOT-OS/RIOT/blob/master/core/lib/init.c#L65) for this?

No, it's no longer mandatory. I haven't updated the original PR description. See the Doxygen Documentation and https://github.com/RIOT-OS/RIOT/pull/17000#issuecomment-1181536070

nmeum commented:
Added a preliminary coverage test application which checks that a coverage dump is properly performed. Ideally, it would also be cool to check the validity of the obtained data by generating gcov files via `dist/tools/coverage/coverage.py` and processing them further using a gcov frontend but that will require additional software etc. so maybe we can add that later.

benpicco commented:
When I run `tests/coverage` on `native` it immediately segfaults. 😕

> `make coverage` still requires the board to be reset by the user.

Could this be solved by adding the `reset` target as a dependency and including the `test_utils_interactive_sync` module so output only starts when the host script gives the start command?

nmeum commented:
> When I run `tests/coverage` on `native` it immediately segfaults.

This is not intended to be run on `native`, on `native` you can directly use gcov instrumentation to determine coverage. Still interesting that it segfaults though, what does the backtrace look like?

> > `make coverage` still requires the board to be reset by the user.
> 
> Could this be solved by adding the `reset` target as a dependency and including the `test_utils_interactive_sync` module so output only starts when the host script gives the start command?

Sounds promising, I will look into it.
